deployment/common/changeset/link_transfer_timelock.go

```diff
@@ -0,0 +1,177 @@
+	"github.com/smartcontractkit/chainlink/deployment/common/types"
+)
+
+type LinkTransfer struct {
+	To    common.Address
+	Value *big.Int
+}
+type LinkTransferTimelockConfig struct {
+	Transfers  map[uint64][]LinkTransfer
+	UseMCMS    bool
+	McmsConfig *MCMSConfig
+}
+type MCMSConfig struct {
+	ValidUntil   uint32        // unix time until the proposal will be valid
+	MinDelay     time.Duration // delay for timelock worker to execute the transfers.
+	OverrideRoot bool
+}
+
+var _ deployment.ChangeSet[*LinkTransferTimelockConfig] = LinkTransferTimelock
+
+// Validate checks that the LinkTransferTimelockConfig is valid.
+func (cfg LinkTransferTimelockConfig) Validate() error {
+
+	// Check that Transfers map has at least one key
+	if len(cfg.Transfers) == 0 {
+		return fmt.Errorf("transfers map must have at least one key")
+	}
+
+	// Check transfers config values.
+	for key, transfers := range cfg.Transfers {
+		if len(transfers) == 0 {
+			return fmt.Errorf("transfers for key %d must have at least one LinkTransfer", key)
+		}
+		for _, transfer := range transfers {
+			if transfer.To == (common.Address{}) {
+				return fmt.Errorf("'to' address for transfers  must be set")
+			}
+			if transfer.Value == nil {
+				return fmt.Errorf("value for transfers must be set")
+			}
+			if transfer.Value == big.NewInt(0) {
+				return fmt.Errorf("value for transfers must be non-zero")
+			}
+		}
+	}
+	// Check that Transfers and StartingOpCount have the same keys
+	for key := range cfg.Transfers {
+		_, err := chain_selectors.GetSelectorFamily(key)
+		if err != nil {
+			return fmt.Errorf("invalid chain selector: %w", err)
+		}
+	}
+	if !cfg.UseMCMS {
+		return nil
+	}
+	// Mcms specific configs
+	if cfg.McmsConfig == nil {
+		return fmt.Errorf("mcmsConfig must be set when UseMCMS is true")
+	}
+	// Upper bound for min delay (7 days)
+	if cfg.McmsConfig.MinDelay > 24*7*time.Hour {
+		return fmt.Errorf("minDelay must be less than 7 days")
+	}
+	return nil
+}
+
+// LinkTransferTimelock takes the given link transfers and executes them or creates an MCMS proposal for them.
+func LinkTransferTimelock(e deployment.Environment, req *LinkTransferTimelockConfig) (deployment.ChangesetOutput, error) {
+	err := req.Validate()
+	ctx := e.GetContext()
+	if err != nil {
+		return deployment.ChangesetOutput{}, fmt.Errorf("invalid LinkTransferTimelockConfig: %w", err)
+	}
+	chainSelectors := []uint64{}
+	for chainSelector := range req.Transfers {
+		chainSelectors = append(chainSelectors, chainSelector)
+	}
+	mcmsPerChain := map[uint64]*owner_helpers.ManyChainMultiSig{}
+
+	timelockAddresses := map[mcms.ChainIdentifier]common.Address{}
+	allBatches := []timelock.BatchChainOperation{}
+	for chainSelector := range req.Transfers {
+		chainID := mcms.ChainIdentifier(chainSelector)
+		chain := e.Chains[chainSelector]
+		addrs, err := e.ExistingAddresses.AddressesForChain(chainSelector)
+		linkState, err := MaybeLoadLinkTokenState(chain, addrs)
```

Comment:
I think you want to load all of these before you start looping over the transfers to prevent partial failures. Probably worth renaming MaybeLoadLinkTokenState to MaybeLoadLinkTokenChainState then adding a multi-chain version MaybeLoadLinkTokenState which just loads link tokens for all chains

Reply:
added an `initStatePerChain`func here that does that here https://github.com/smartcontractkit/chainlink/pull/15512/commits/f889dfc35991573e31c5a56015b3780ff673a9a5

Reply:
and here im renaming the load state funcs https://github.com/smartcontractkit/chainlink/pull/15512/commits/3d2dad126d7309dbc9f1beb1971fa0670a5aaadb